python/ray/air/_internal/usage.py

```diff
@@ -18,16 +20,106 @@
+# This is the best we can do for now.
+FLAML_SEARCHERS = {"FLOW2"}
+
+FLAML_SEARCHER_MODULE_PATH = "flaml.tune.searcher"
+
+TUNE_SEARCHER_WRAPPER = {
+    "ConcurrencyLimiter",
+    "Repeater",
+}
+
+TUNE_SCHEDULERS = {
+    "FIFOScheduler",
+    "AsyncHyperBandScheduler",
+    "AsyncHyperBandScheduler",
+    "MedianStoppingRule",
+    "HyperBandScheduler",
+    "HyperBandForBOHB",
+    "PopulationBasedTraining",
+    "PopulationBasedTrainingReplay",
+    "PB2",
+    "ResourceChangingScheduler",
+}
+
+
+def _find_class_name(obj, allowed_module_path_prefix: str, whitelist: Set[str]):
+    """Find the class name of the object. If the object is not
+    under `allowed_module_path_prefix` or if its class is not in the whitelist,
+    return "Custom".
+
+    Args:
+        obj: The object under inspection.
+        allowed_module_path_prefix: If the `obj`'s class is not under
+            the `allowed_module_path_prefix`, its class name will be anonymized.
+        whitelist: If the `obj`'s class is not in the `whitelist`,
+            it will be anonymized.
+    Returns:
+        The class name to be tagged with telemetry.
+    """
+    module_path = obj.__module__
+    if module_path.startswith(allowed_module_path_prefix):
+        cls_name = obj.__class__.__name__
+        return cls_name if cls_name in whitelist else "Custom"
+    else:
+        return "Custom"
+
 
 def tag_air_trainer(trainer: "BaseTrainer"):
     from ray.train.trainer import BaseTrainer
 
     assert isinstance(trainer, BaseTrainer)
-    module_path = trainer.__module__
-    if module_path.startswith("ray.train"):
-        trainer_name = trainer.__class__.__name__
-        if trainer_name not in AIR_TRAINERS:
-            trainer_name = "Custom"
-    else:
-        trainer_name = "Custom"
+    trainer_name = _find_class_name(trainer, "ray.train", AIR_TRAINERS)
     record_extra_usage_tag(TagKey.AIR_TRAINER, trainer_name)
+
+
+def tag_searcher(searcher: Union["BasicVariantGenerator", "Searcher"]):
+    from ray.tune.search import BasicVariantGenerator, Searcher
+
+    if isinstance(searcher, BasicVariantGenerator):
+        # Note this could be highly inflated as all train flows are treated
+        # as using BasicVariantGenerator.
+        record_extra_usage_tag(TagKey.TUNE_SEARCHER, "BasicVariantGenerator")
+    elif isinstance(searcher, Searcher):
+        searcher_name = _find_class_name(
+            searcher, "ray.tune.search", TUNE_SEARCHERS.union(TUNE_SEARCHER_WRAPPER)
+        )
+        if searcher_name in TUNE_SEARCHER_WRAPPER:
+            # ignore to avoid double tagging with wrapper name.
+            return
+        if searcher_name == "Custom":
+            # try one more time to see if flaml searchers
+            searcher_name = _find_class_name(
+                searcher, FLAML_SEARCHER_MODULE_PATH, FLAML_SEARCHERS
+            )
+        record_extra_usage_tag(TagKey.TUNE_SEARCHER, searcher_name)
```

Comment:
I wonder if we even need the special logic to check the module path. For example, if we were to just whitelist the class name that we know and share in the documentation, e.g. `FLOW2` or `OptunaSearch`, does it really matter if the module is specifically `ray.tune.search` or some other module?

Reply:
And what does that gain us? I feel like this is a very shady path we are going down anyways. For example, whether `FLOW2` shows up or some other name shows up purely relies on the implementation of this file [`blendsearch.py`](https://github.com/microsoft/FLAML/blob/main/flaml/tune/searcher/blendsearch.py#L207). They may change it at any given time.

Reply:
Actually a less shady approach would be to tag it within *our* `flaml.__init__` :) let me know if you are ok with it. We may just have to special handle it a bit...

Reply:
Yeah... my intention was to try to reduce the special handling as much as possible, but not sure how effective my proposal would actually be here.

For the `init` problem, I actually do think in the long run it would be cleaner to call these `record` utils in some central execution path (e.g. in `tune.run`), similar to what was done in your original PR. Overall it would look similar to some validation step of all the Tune arguments. Though this may be out of scope for this initial PR since it would involve even more custom logic to handle things like `ConcurrencyLimiter`/`Repeater`.

I'd generally be okay to tag in our `flaml.__init__` logic if it doesn't end up too hacky. I'm also okay to just start with searchers that are defined exclusively in `ray` and extend for these aliased implementations in a separate PR.

Reply:
copying over offline chat:
* for *our* searchers let’s do it in `Searcher.__init__` and since we keep a whitelist anyways, we will not log some external usage unexpectedly. :sweat_smile:
* for flaml (which is like the only special one right now), let’s just do it in `flaml.__init__` in our code.